compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs

```diff
@@ -70,120 +72,135 @@ import SdkVersion
-
-    -- Iterate over the dependency graph in topological order.
-    -- We do a topological sort on the transposed graph which ensures that
-    -- the packages with no dependencies come first and we
-    -- never process a package without first having processed its dependencies.
-    forM_ (topSort $ transposeG depGraph) $ \vertex ->
-      let (pkgNode, pkgId) = vertexToNode vertex in
-      -- stable packages are mapped to the current version of daml-prim/daml-stdlib
-      -- so we don’t need to generate interface files for them.
-      unless (pkgId `Set.member` stablePkgIds || pkgId `Set.member` dependenciesInPkgDbIds) $ do
-        let unitIdStr = unitIdString $ unitId pkgNode
-        let pkgIdStr = T.unpack $ LF.unPackageId pkgId
-        let (pkgName, mbPkgVersion) = LF.splitUnitId (unitId pkgNode)
-        let deps =
-                [ (unitId depPkgNode, dalfPackage depPkgNode)
-                | (depPkgNode, depPkgId) <- map vertexToNode $ reachable depGraph vertex
-                , pkgId /= depPkgId
-                , not (depPkgId `Set.member` stablePkgIds)
-                ]
-        let workDir = dbPath </> unitIdStr <> "-" <> pkgIdStr
-        createDirectoryIfMissing True workDir
-        -- write the dalf package
-        BS.writeFile (workDir </> unitIdStr <.> "dalf") $ LF.dalfPackageBytes (dalfPackage pkgNode)
-
-        generateAndInstallIfaceFiles
-            (LF.extPackagePkg $ LF.dalfPackagePkg $ dalfPackage pkgNode)
-            (stubSources pkgNode)
-            opts
-            workDir
-            dbPath
-            projectPackageDatabase
-            pkgId
-            pkgName
-            mbPkgVersion
-            deps
-            dependenciesInPkgDb
-            exposedModules
-
-    writeMetadata projectRoot (PackageDbMetadata (mainUnitIds dependencyInfo) validatedModulePrefixes)
+createProjectPackageDb projectRoot (disableScenarioService -> opts) (PackageSdkVersion thisSdkVer) modulePrefixes deps' dataDeps
+  = do
+    -- compute the hash over all dependencies
+    deps <- expandSdkPackages (optDamlLfVersion opts) (filter (`notElem` basePackages) deps')
+    depHash <-
+        fmap hash $
+        foldM
+            (\acc dep -> do
+                 bs <- BSL.readFile dep
+                 pure $ encode $ hash $ bs <> acc)
+            BSL.empty
+            (deps ++ dataDeps)
+
+    -- Read the metadata of an already existing package database and see if wee need to reinitialize.
+    let metadataFileFp = metadataFile projectRoot
+    hasMetadata <- doesFileExist metadataFileFp
+    needsReinitalization <- if hasMetadata
+                            then do
+                              metaData <- readMetadata projectRoot
+                              pure $ hashDependencies metaData /= Just depHash
+                            else pure True
+    when needsReinitalization $ do
```

Comment:
This is growing quite long, maybe a good idea to factor out the check as well as the body of the `when`, so this just turns into something like
```
createProjectPackageDb … = do
  needsReinitialization <- getNeedsReinitialization
  when needsReinitialization $ initializeDb
```

Fine to leave that for a separate PR.

Reply:
I'll factor it out.